File: 0714. Best Time to Buy and Sell Stock with Transaction Fee/Solution0714.js
/**
 * @param {number[]} prices
 * @param {number} fee
 * @return {number}
 */
var maxProfit = function(prices, fee) {
    let n = prices.length, buy = -prices[0], sell = 0;
    for (let i = 1; i < n; i++) {
        buy = Math.max(buy, sell - prices[i]);
        sell = Math.max(sell, buy + prices[i] - fee);
    }
    return sell;    
};
var prices = [1, 3, 2, 8, 4, 9];
console.log(prices);
var res = maxProfit(prices,2);
console.log(res);

File: 0714. Best Time to Buy and Sell Stock with Transaction Fee/Solution0714.go
package main

import (
	"fmt"
)

func main() {
	arr := []int{1, 3, 2, 8, 4, 9}
	fee := 2
	res := maxProfit(arr, fee)
	fmt.Println(res)
}

func maxProfit(prices []int, fee int) int {
	buy := -prices[0]
	sell := 0
	for i := 1; i < len(prices); i++ {
		buy = max(buy, sell-prices[i])
		sell = max(sell, buy+prices[i]-fee)
	}
	return sell
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
